contracts/ethregistrar/ETHRegistrarController.sol

```diff
@@ -1,178 +1,275 @@
 pragma solidity >=0.8.4;
 
-import "./PriceOracle.sol";
 import "./BaseRegistrarImplementation.sol";
 import "./StringUtils.sol";
-import "@openzeppelin/contracts/access/Ownable.sol";
 import "../resolvers/Resolver.sol";
+import "../registry/ReverseRegistrar.sol";
+import "../dnssec-oracle/BytesUtils.sol";
+import "./IETHRegistrarController.sol";
+
+import "@openzeppelin/contracts/access/Ownable.sol";
+import "@openzeppelin/contracts/utils/introspection/IERC165.sol";
+import "../wrapper/INameWrapper.sol";
 
 /**
  * @dev A registrar controller for registering and renewing names at fixed cost.
  */
-contract ETHRegistrarController is Ownable {
+contract ETHRegistrarController is Ownable, IETHRegistrarController {
     using StringUtils for *;
-
-    uint constant public MIN_REGISTRATION_DURATION = 28 days;
-
-    bytes4 constant private INTERFACE_META_ID = bytes4(keccak256("supportsInterface(bytes4)"));
-    bytes4 constant private COMMITMENT_CONTROLLER_ID = bytes4(
-        keccak256("rentPrice(string,uint256)") ^
-        keccak256("available(string)") ^
-        keccak256("makeCommitment(string,address,bytes32)") ^
-        keccak256("commit(bytes32)") ^
-        keccak256("register(string,address,uint256,bytes32)") ^
-        keccak256("renew(string,uint256)")
+    using BytesUtils for bytes;
+
+    uint256 public constant MIN_REGISTRATION_DURATION = 28 days;
+    bytes32 private constant ETH_NODE =
+        0x93cdeb708b7545dc668eb9280176169d1c33cfd8ed6f04690a0bcc88a93fc4ae;
+
+    BaseRegistrarImplementation immutable base;
+    PriceOracle public prices;
```

Comment:
Where did we come down on making this immutable? Deploying a new controller for each oracle change would be a pain, but it would cut a bit of gas off every registration tx.

Reply:
I think we can make it immutable. I don't think we'll end up changing the oracle often if we can sort out this exponential oracle out before all these contracts are ready